src/snowflake/cli/_plugins/nativeapp/sf_sql_facade.py

```diff
@@ -0,0 +1,224 @@
+# Copyright (c) 2024 Snowflake Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+from contextlib import contextmanager
+
+from click import ClickException
+from cryptography.utils import cached_property
+from snowflake.cli.api.constants import ObjectType
+from snowflake.cli.api.entities.common import get_sql_executor
+from snowflake.cli.api.errno import (
+    DOES_NOT_EXIST_OR_CANNOT_BE_PERFORMED,
+    NO_WAREHOUSE_SELECTED_IN_SESSION,
+)
+from snowflake.cli.api.exceptions import CouldNotUseObjectError
+from snowflake.cli.api.project.util import to_identifier
+from snowflake.cli.api.sql_execution import SqlExecutor
+from snowflake.connector import ProgrammingError
+
+
+class UnknownSQLError(Exception):
+    """Exception raised when the root of the SQL error is unidentified by us."""
+
+    # PJ-question how do we ensure exit codes remain unique
+    exit_code = 3
```

Comment:
any thoughts/insights on this? cc @sfc-gh-fcampbell

Reply:
i've done something like this before:
```python
class UniqueExitCode(type):
	_exit_code = 3
	
	def __new__(mcs, name, bases, attrs):
		c = mcs._exit_code
		if c > 255:
			raise ValueError(f"too many uses of {mcs.__name__}, max exit code reached")
		mcs._exit_code += 1
		attrs["exit_code"] = c
		return type(name, bases, attrs)

class MyError1(Exception, metaclass=UniqueExitCode):
	pass

class MyError2(Exception, metaclass=UniqueExitCode):
	pass
```

Reply:
Two questions:
1. Is there potentially a race condition where the exit codes for errors could change depending on when the code is called, refactors, and new errors that are introduced?
2. What is the purpose of making exit codes unique here? I'm assuming it's to identify the type of error, but it doesn't seem like there's any relationship between the exit code and the error type since they all fall under `UnknownSQLError`?

Reply:
good point - no need for it to be unique. I had the wrong idea of what we might be using it for.